File: app/src/main/java/com/cengels/progressor/units/UnitValue.java
package com.cengels.progressor.units;

import android.annotation.SuppressLint;
import androidx.annotation.NonNull;
import com.cengels.progressor.enums.ProgressType;

import java.io.Serializable;
import java.util.Objects;

public class UnitValue implements Serializable {
    public static final int SECS_PER_MINUTE = 60;
    private double value;
    @NonNull
    private final ProgressType type;
    private int decimals = -1;
    private String unit;

    public UnitValue(@NonNull final ProgressType type) {
        this.type = type;
        this.unit = Unit.getSmallestUnit(this.type);

        if (this.unit == null) {
            this.unit = "";
        }

        this.inferDecimals();
    }

    public UnitValue(final double value, @NonNull final ProgressType type) {
        this(type);
        this.value = value;
    }

    public UnitValue(@NonNull final String unit, final int decimals) {
        this.type = Unit.getProgressType(unit);
        this.unit = unit;
        this.decimals = decimals;
    }

    public UnitValue(final double value, @NonNull final String unit) {
        this(unit, 2);
        this.value = value;

        this.inferDecimals();
    }

    public UnitValue(final int value, @NonNull final String unit) {
        this(unit, 0);
        this.value = value;

        this.inferDecimals();
    }

    public UnitValue(final double value, @NonNull final String unit, final int decimals) {
        this(unit, decimals);
        this.value = value;
    }

    public UnitValue(final int value, @NonNull final String unit, final int decimals) {
        this(unit, decimals);
        this.value = value;
    }

    public void set(final double value) {
        this.value = value;
    }

    public double get() {
        return this.value;
    }

    @NonNull
    public ProgressType getType() {
        return this.type;
    }

    public int getDecimals() {
        return this.decimals;
    }

    public String getUnit() {
        return this.unit;
    }

    public String getBestUnit() {
        if (this.type == ProgressType.CUSTOM) {
            return this.unit;
        }

        return Convert.from(this).best().getUnit();
    }

    public UnitValue getBest() {
        if (this.type == ProgressType.CUSTOM) {
            return this;
        }

        return Convert.from(this).best();
    }

    private void inferDecimals() {
        switch (this.type) {
            case COUNT:
            case TIME:
                this.decimals = 0;
                break;
            case PERCENTAGE:
                this.decimals = 1;
                break;
            default:
                this.decimals = 2;
                break;
        }
    }

    @SuppressLint("DefaultLocale")
    public String getFormattedValue() {
        final UnitValue value = this.getBest();

        switch (value.getUnit()) {
            case Unit.SECONDS:
                return String.format("0:%02d", (long)value.get());
            case Unit.MINUTES:
                return String.format("%d:%02d", (long)value.get(), (long)(value.get() * SECS_PER_MINUTE % SECS_PER_MINUTE));
            case Unit.HOURS:
                return String.format("%d:%02d:%02d", (long)value.get(), (long)(value.get() * SECS_PER_MINUTE % SECS_PER_MINUTE), (long)(value.get() * SECS_PER_MINUTE * SECS_PER_MINUTE % SECS_PER_MINUTE));
        }

        if (value.decimals == 0 || (long)value.get() == value.get()) {
            return String.valueOf((long)value.get());
        }

        return String.format("%." + this.decimals + "f", value.get());
    }

    public String getFormattedValue(@NonNull final String unit) {
        final UnitValue value = Convert.from(this).to(unit);

        if (this.decimals == 0) {
            return String.valueOf((long)value.get());
        }

        return String.format("%." + this.decimals + "f", value.get());
    }

    @Override
    public String toString() {
        final UnitValue best = this.getBest();

        switch (best.getUnit()) {
            case Unit.SECONDS:
            case Unit.MINUTES:
            case Unit.HOURS:
                return best.getFormattedValue();
        }

        return best.getFormattedValue() + best.getUnit();
    }

    public String toString(@NonNull final String unit) {
        switch (unit) {
            case Unit.SECONDS:
            case Unit.MINUTES:
            case Unit.HOURS:
                return this.getFormattedValue(unit);
        }

        return this.getFormattedValue(unit) + unit;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }

        if (o == null || this.getClass() != o.getClass()) {
            return false;
        }

        final UnitValue unitValue = (UnitValue) o;
        return Double.compare(unitValue.value, this.value) == 0 &&
                this.decimals == unitValue.decimals &&
                this.type == unitValue.type &&
                this.unit.equals(unitValue.unit);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.value, this.type, this.decimals, this.unit);
    }
}

File: app/src/main/kotlin/com/cengels/progressor/units/UnitValue.kt
package com.cengels.progressor.units

import com.cengels.progressor.enums.ProgressType

import java.io.Serializable
import kotlin.math.roundToInt

data class UnitValue private constructor(
        private val value: Double,
        var unit: String,
        var decimals: Int = 2,
        val type: ProgressType
) : Serializable {
    fun getBestUnit(): String? = if (this.type == ProgressType.CUSTOM) this.unit else Convert.from(this).best().unit

    fun getBest(): UnitValue = if (this.type == ProgressType.CUSTOM) this else Convert.from(this).best()

    val formattedValue: String
        get() {
            val value = this.getBest()

            when (value.unit) {
                Unit.SECONDS -> return String.format("0:%02d", value.get().toLong())
                Unit.MINUTES -> return String.format("%d:%02d", value.get().toLong(), (value.get() * SECS_PER_MINUTE % SECS_PER_MINUTE).toLong())
                Unit.HOURS -> return String.format("%d:%02d:%02d", value.get().toLong(), (value.get() * SECS_PER_MINUTE % SECS_PER_MINUTE).toLong(), (value.get() * SECS_PER_MINUTE.toDouble() * SECS_PER_MINUTE.toDouble() % SECS_PER_MINUTE).toLong())
            }

            return if (value.decimals == 0 || value.get().toLong().toDouble() == value.get()) {
                value.get().toLong().toString()
            } else String.format("%." + this.decimals + "f", value.get())

        }

    constructor(value: Double, unit: String, decimals: Int) : this(value, unit, decimals, Unit.getProgressType(unit))
    constructor(value: Double, unit: String) : this(value, unit, -1, Unit.getProgressType(unit)) {
        this.inferDecimals()
    }
    constructor(value: Double, type: ProgressType) : this(value, Unit.getSmallestUnit(type) ?: "", 2, type) {
        this.inferDecimals()
    }

    constructor(value: Int, unit: String, decimals: Int) : this(value.toDouble(), unit, decimals)
    constructor(value: Int, unit: String) : this(value.toDouble(), unit)

    constructor(type: ProgressType) : this(0.0, Unit.getSmallestUnit(type) ?: "", -1, type) {
        this.inferDecimals()
    }
    constructor(unit: String, decimals: Int) : this(0.0, unit, decimals, Unit.getProgressType(unit))

    /** Gets the raw value contained within this [UnitValue]. */
    fun get(): Double = this.value
    /** Gets a rounded int value from the raw value contained within this [UnitValue]. */
    fun getInt(): Int = this.value.roundToInt()

    private fun inferDecimals() {
        when (this.type) {
            ProgressType.COUNT, ProgressType.TIME -> this.decimals = 0
            ProgressType.PERCENTAGE -> this.decimals = 1
            else -> this.decimals = 2
        }
    }

    /** Gets a string representation of the value contained within this [UnitValue]. */
    fun getFormattedValue(unit: String? = null): String {
        val value = if (unit != null) Convert.from(this).to(toUnit = unit) else Convert.from(this).best()

        return if (this.decimals == 0) value.get().toLong().toString() else String.format("%.${this.decimals}f", value.get())
    }

    override fun toString(): String {
        return when (this.getBest().unit) {
            Unit.SECONDS, Unit.MINUTES, Unit.HOURS -> getBest().formattedValue
            else -> getBest().formattedValue + getBest().unit
        }
    }

    fun toString(unit: String): String {
        return when (unit) {
            Unit.SECONDS, Unit.MINUTES, Unit.HOURS -> this.getFormattedValue(unit)
            else -> this.getFormattedValue(unit) + unit
        }
    }

    operator fun times(value: UnitValue): UnitValue = UnitValue(this.value * value.value, this.unit, this.decimals)
    operator fun div(value: UnitValue): UnitValue = UnitValue(this.value / value.value, this.unit, this.decimals)
    operator fun plus(value: UnitValue): UnitValue = UnitValue(this.value + value.value, this.unit, this.decimals)
    operator fun minus(value: UnitValue): UnitValue = UnitValue(this.value - value.value, this.unit, this.decimals)

    operator fun times(value: Double): UnitValue = UnitValue(this.value * value, this.unit, this.decimals)
    operator fun div(value: Double): UnitValue = UnitValue(this.value / value, this.unit, this.decimals)
    operator fun plus(value: Double): UnitValue = UnitValue(this.value + value, this.unit, this.decimals)
    operator fun minus(value: Double): UnitValue = UnitValue(this.value - value, this.unit, this.decimals)

    operator fun times(value: Int): UnitValue = UnitValue(this.value * value, this.unit, this.decimals)
    operator fun div(value: Int): UnitValue = UnitValue(this.value / value, this.unit, this.decimals)
    operator fun plus(value: Int): UnitValue = UnitValue(this.value + value, this.unit, this.decimals)
    operator fun minus(value: Int): UnitValue = UnitValue(this.value - value, this.unit, this.decimals)

    operator fun inc(): UnitValue = this + 1

    companion object {
        const val SECS_PER_MINUTE: Int = 60
    }
}
